rpc/json-rpc/src/error/code.rs

```diff
@@ -0,0 +1,224 @@
+    /// Invalid JSON was received by the server.
+    ///
+    /// An error occurred on the server while parsing the JSON text.
+    ParseError,
+
+    #[error("{}: {}", INVALID_REQUEST.0, INVALID_REQUEST.1)]
+    /// The JSON sent is not a valid Request object.
+    InvalidRequest,
+
+    #[error("{}: {}", METHOD_NOT_FOUND.0, METHOD_NOT_FOUND.1)]
+    /// The method does not exist / is not available.
+    MethodNotFound,
+
+    #[error("{}: {}", INVALID_PARAMS.0, INVALID_PARAMS.1)]
+    /// Invalid method parameters.
+    InvalidParams,
+
+    #[error("{}: {}", INTERNAL_ERROR.0, INTERNAL_ERROR.1)]
+    /// Internal JSON-RPC error.
+    InternalError,
+
+    #[error("{0}: {SERVER_ERROR}")]
+    /// Reserved for implementation-defined server-errors.
+    ServerError(i32),
+}
+
+impl ErrorCode {
+    /// Creates [`Self`] from a [`i32`] code.
+    ///
+    /// [`From<i32>`] is the same as this function.
+    ///
+    /// ```rust
+    /// use json_rpc::error::{
+    ///     ErrorCode,
+    ///     INTERNAL_ERROR, INVALID_PARAMS, INVALID_REQUEST, METHOD_NOT_FOUND, PARSE_ERROR,
+    /// };
+    ///
+    /// assert_eq!(ErrorCode::from_code(PARSE_ERROR.0),      ErrorCode::ParseError);
+    /// assert_eq!(ErrorCode::from_code(INVALID_REQUEST.0),  ErrorCode::InvalidRequest);
+    /// assert_eq!(ErrorCode::from_code(METHOD_NOT_FOUND.0), ErrorCode::MethodNotFound);
+    /// assert_eq!(ErrorCode::from_code(INVALID_PARAMS.0),   ErrorCode::InvalidParams);
+    /// assert_eq!(ErrorCode::from_code(INTERNAL_ERROR.0),   ErrorCode::InternalError);
+    ///
+    /// // Non-defined code inputs will default to a custom `ServerError`.
+    /// assert_eq!(ErrorCode::from_code(0), ErrorCode::ServerError(0));
+    /// assert_eq!(ErrorCode::from_code(1), ErrorCode::ServerError(1));
+    /// assert_eq!(ErrorCode::from_code(2), ErrorCode::ServerError(2));
+    /// ```
+    pub const fn from_code(code: i32) -> Self {
+        /// HACK: you cannot `match` on tuple fields
+        /// like `PARSE_ERROR.0 => /*...*/` so extract
+        /// it out here.
+        #[allow(clippy::wildcard_imports, clippy::missing_docs_in_private_items)]
+        mod i32s {
+            use super::*;
+            pub(super) const PARSE_ERROR_I32: i32 = PARSE_ERROR.0;
+            pub(super) const INVALID_REQUEST_I32: i32 = INVALID_REQUEST.0;
+            pub(super) const METHOD_NOT_FOUND_I32: i32 = METHOD_NOT_FOUND.0;
+            pub(super) const INVALID_PARAMS_I32: i32 = INVALID_PARAMS.0;
+            pub(super) const INTERNAL_ERROR_I32: i32 = INTERNAL_ERROR.0;
+        }
+        #[allow(clippy::wildcard_imports)]
+        use i32s::*;
+
+        match code {
+            PARSE_ERROR_I32 => Self::ParseError,
+            INVALID_REQUEST_I32 => Self::InvalidRequest,
+            METHOD_NOT_FOUND_I32 => Self::MethodNotFound,
+            INVALID_PARAMS_I32 => Self::InvalidParams,
+            INTERNAL_ERROR_I32 => Self::InternalError,
+            code => Self::ServerError(code),
+        }
+    }
```

Comment:
Better way to do this?

Reply:
I can't think of one, this is very ugly though

Reply:
Replaced with `if` chain, seems to compile to the same assembly.

Reply:
I prefer seeing this syntax than an `if` chain tbh